### PR TITLE
web and nginx slaves

### PR DESCRIPTION
**Pull request by atarkowska**
This PR improves existing web infrastructure introduced in https://github.com/openmicroscopy/devspace/pull/19

```
$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                                   NAMES
2036d9641125        devspace_web        "/usr/sbin/init"         2 minutes ago       Up 2 minutes        22/tcp, 0.0.0.0:4080->4080/tcp          devspace_web_1
5f94f815fe7b        devspace_slave      "/tmp/jenkins-slave.s"   2 minutes ago       Up 2 minutes                                                devspace_slave_1
a773bc06f903        devspace_nginx      "/usr/sbin/init"         2 minutes ago       Up 2 minutes        22/tcp, 443/tcp, 0.0.0.0:8888->80/tcp   devspace_nginx_1
41121d7114bd        devspace_jenkins    "/bin/tini -- /usr/lo"   2 minutes ago       Up 2 minutes        0.0.0.0:8080->8080/tcp, 50000/tcp       devspace_jenkins_1
89b0a8f675b1        postgres            "/docker-entrypoint.s"   2 minutes ago       Up 2 minutes        5432/tcp                                devspace_pg_1
0d64ddcb5ce7        redis               "/entrypoint.sh redis"   2 minutes ago       Up 2 minutes        0.0.0.0:6379->6379/tcp                  devspace_redis_1
```

nginx is available on http://$(docker-machine ip omerodev):8888
- use swarm plugin 2.0
- use latest omero-install and fix Dockerfile
- make nginx a jenkins node
- allow systemd to start services on web and nginx
- allow linking nginx config and statics
- add redis node

```
jenkins_1 | Mar 26, 2016 9:41:49 PM hudson.TcpSlaveAgentListener$ConnectionHandler run
jenkins_1 | INFO: Accepted connection #1 from /172.17.0.7:58020
jenkins_1 | Mar 26, 2016 9:41:49 PM hudson.TcpSlaveAgentListener$ConnectionHandler run
jenkins_1 | INFO: Accepted connection #2 from /172.17.0.6:38142
jenkins_1 | Mar 26, 2016 9:41:49 PM hudson.TcpSlaveAgentListener$ConnectionHandler run
jenkins_1 | INFO: Accepted connection #3 from /172.17.0.5:42540
```
### NOTE:
- https://github.com/openmicroscopy/openmicroscopy/pull/4544 is required

### CODE REVIEW

**Comment by atarkowska** [comment]
@sbesson could you advice why BF job is failing?

```
2016-03-24 19:38:07,333 [   scc.merge] INFO  Merged PRs:
2016-03-24 19:38:07,333 [   scc.merge] INFO    # PR 1828 melissalinkert 'JPEG: add basic support for CMYK images'
...
2016-03-24 19:38:07,334 [   scc.merge] INFO    # PR 2305 melissalinkert 'Zeiss CZI: fixes for files with multiple mosaics and positions'
2016-03-24 19:38:07,334 [   scc.merge] INFO  

/home/slave/.local/lib/python2.6/site-packages/scc/git.py:79: DeprecationWarning: BaseException.message has been deprecated as of Python 2.6
  if exception.message != "rc=128":

Traceback (most recent call last):
  File "/home/slave/.local/lib/python2.6/site-packages/scc/main.py", line 76, in entry_point
    (UpdateSubmodules.NAME, UpdateSubmodules),
  File "/home/slave/.local/lib/python2.6/site-packages/yaclifw/framework.py", line 197, in main
    ns.func(ns)
  File "/home/slave/.local/lib/python2.6/site-packages/scc/git.py", line 2912, in __call__
    self.push(args, self.main_repo)
  File "/home/slave/.local/lib/python2.6/site-packages/scc/git.py", line 1984, in push
    main_repo.rpush(branch_name, remote, force=True)
  File "/home/slave/.local/lib/python2.6/site-packages/scc/git.py", line 1833, in rpush
    self.push_branch(branch_name, remote=full_remote, force=force)
  File "/home/slave/.local/lib/python2.6/site-packages/scc/git.py", line 98, in wrapper
    return func(*args, **kwargs)
  File "/home/slave/.local/lib/python2.6/site-packages/scc/git.py", line 1147, in push_branch
    self.call("git", "push", "-f", remote, name)
  File "/home/slave/.local/lib/python2.6/site-packages/scc/git.py", line 1013, in call
    return self.wrap_call(self.debugWrap, *command, **kwargs)
  File "/home/slave/.local/lib/python2.6/site-packages/scc/git.py", line 1038, in wrap_call
    raise Exception("rc=%s" % rc)
Exception: rc=128
Build step 'Execute shell' marked build as failure

Finished: FAILURE
```

**Comment by sbesson** [comment]
@aleksandra-tarkowska: this looks like a permission issue while pushing the merge branch. Did you set up the SSH keys and `.gitconfig` on the slave workspace as described in the README?

**Comment by atarkowska** [comment]
I could forget but after coping rebuilding is the same

File: web/jenkins.service [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+Description=Jenkins slave
+
+[Service]
+User=omero
+WorkingDirectory=/tmp
+PIDFile=/tmp/jenkins.pid
+Type=forking
+Restart=on-failure
+RestartSec=10
+ExecStart=/tmp/jenkins-slave.sh
```
Comment (manics) [inline web/jenkins.service]: Does this start jenkins as a background (daemon) process, or in the foreground? If it's in the foreground you can use Type `simple` instead and omit the `ExecStop`, e.g. see https://github.com/ome/omero-install/blob/develop/linux/omero-systemd.service

Comment (atarkowska) [inline web/jenkins.service]: you mean https://documentation.cloudbees.com/docs/dev-at-cloud/On-Premise+Executors.html?query=systemd#running-the-slave-with-systemd-on-linux
???

```
[root@102787622c66 ~]# systemctl status jenkins.service
● jenkins.service - Jenkins slave
   Loaded: loaded (/etc/systemd/system/jenkins.service; enabled; vendor preset: disabled)
   Active: active (running) since Fri 2016-03-25 16:04:12 UTC; 1min 14s ago
     Docs: https://developer.cloudbees.com/bin/view/DEV/On-Premise+Executors
 Main PID: 216 (java)
   CGroup: /system.slice/jenkins.service
           └─216 java -jar /tmp/swarm-client-2.0-jar-with-dependencies.jar -fsroot /home/omero -labels centos7 -labels java -name web -disableClientsUniq...

Mar 25 16:05:03 102787622c66 jenkins-slave.sh[216]: Retrying in 10 seconds
Mar 25 16:05:16 102787622c66 systemd[1]: Started Jenkins slave.
Mar 25 16:05:18 102787622c66 jenkins-slave.sh[216]: Failed to receive a reply to broadcast.
```

Comment (manics) [inline web/jenkins.service]: Yes

File: nginx/jenkins-slave.sh [hunk 2]
```diff
@@ -1,6 +1,8 @@
 #!/bin/bash
 # https://raw.githubusercontent.com/carlossg/jenkins-swarm-slave-docker/master/jenkins-slave.sh
 
+eval `(sudo cat /proc/1/environ; echo) | tr '\000' '\n' | while read line; do if [[ $line == JENKINS_PORT_8080_TCP_* ]] ; then echo export $line; fi done`
```
Comment (atarkowska) [inline nginx/jenkins-slave.sh:4]: @manics indeed dirty hack works ;-)

**Comment by atarkowska** [comment]
Could anyone suggest why am I getting:

```
Building remotely on slave (java swarm) in workspace /home/slave/workspace/OMERO-push
Installing JDK jdk-8u74-oth-JPR
Downloading JDK from http://download.oracle.com/otn/java/jdk/8u74-b02/jdk-8u74-linux-x64.tar.gz
Your Oracle account doesn't appear valid. Please specify a valid username/password
ERROR: Unable to install JDK unless a valid username/password is provided.
Finished: FAILURE
```

File: docker-compose.yml [hunk 3]
```diff
@@ -21,32 +21,52 @@ slave:
...
+    command: -labels java -labels slave -name slave -disableClientsUniqueId -executors 4
 
 web:
     build: web
-    environment:
-        - JENKINS_LABEL=/etc/alternatives/java_sdk
-        - JENKINS_MODE=exclusive
-        - SERVICE_NAME=SPACENAME-omero
+    ports:
+        - "4080:4080"
```
Comment (joshmoore) [inline docker-compose.yml:37]: This will need to be in the overwrite file (e.g. docker-compose.osx.yml) since otherwise only one devspace can run on a single docker host.

File: docker-compose.yml [hunk 4]
```diff
@@ -21,32 +21,52 @@ slave:
...
+    environment:
+        - JENKINS_LABEL=/etc/alternatives/java_sdk
+        - JENKINS_MODE=exclusive
+        - SERVICE_NAME=SPACENAME-web
 
 nginx:
-    image: nginx
+    build: nginx
+    ports:
+        - "8888:80"
```
Comment (joshmoore) [inline docker-compose.yml:55]: See https://github.com/joshmoore/devspace/pull/1/files#r57531628

File: docker-compose.yml [hunk 5]
```diff
@@ -21,32 +21,52 @@ slave:
...
     environment:
-        - SERVICE_NAME=SPACENAME-web
+        - JENKINS_LABEL=/etc/alternatives/java_sdk
+        - JENKINS_MODE=exclusive
+        - SERVICE_NAME=SPACENAME-nginx
+
+redis:
+    image: redis
+    ports:
+        - "6379:6379"
```
Comment (joshmoore) [inline docker-compose.yml:70]: See https://github.com/joshmoore/devspace/pull/1/files#r57531628

File: home/jobs/OMERO-build/config.xml [hunk 6]
```diff
@@ -13,7 +13,7 @@
     </userRemoteConfigs>
     <branches>
       <hudson.plugins.git.BranchSpec>
-        <name>SPACEBRANCH/merge/trigger</name>
+        <name>SPACEBRANCH/merge/daily</name>
```
Comment (joshmoore) [inline home/jobs/OMERO-build/config.xml:16]: Why the change here?

File: home/jobs/OMERO-web/config.xml [hunk 7]
```diff
@@ -32,22 +32,22 @@
...
+OMERO_DIST=$HOME/OMERO.server
 if [ -e $OMERO_DIST ]; then
-
-	if [ -e $WORKSPACE/omeropy-virtualenv ]; then
-		source $WORKSPACE/omeropy-virtualenv/bin/activate
+	if [ -e $HOME/omero-virtualenv ]; then
+		source $HOME/omero-virtualenv/bin/activate
         eval $(bash /opt/multi-config.sh ice3.5)
-    	$OMERO_DIST/bin/omero web stop
+        sudo systemctl stop omero-web.service
```
Comment (joshmoore) [inline home/jobs/OMERO-web/config.xml:41]: No objections here, but this couples us tightly to C7. Any way to do this so that if we move to another platform it will still work?

File: home/jobs/OMERO-web/config.xml [hunk 8]
```diff
@@ -68,24 +69,56 @@ unzip $ZIP
...
+$OMERO_DIST/bin/omero config set omero.web.prefix &apos;/web&apos;
+$OMERO_DIST/bin/omero config set omero.web.static_url &apos;/web/static/&apos;
+$OMERO_DIST/bin/omero config set omero.web.static_root &apos;/home/omero/nginx/web&apos;
 
-sudo -u root bash step05_centos7_nginx.sh
-sudo -u root bash step06_centos7_daemon.sh
+$OMERO_DIST/bin/omero config set omero.web.session_engine &apos;django.contrib.sessions.backends.cache&apos;
+$OMERO_DIST/bin/omero config set omero.web.caches &apos;{&quot;default&quot;: {&quot;BACKEND&quot;: &quot;redis_cache.RedisCache&quot;,&quot;LOCATION&quot;: &quot;192.168.99.100:6379&quot;}}&apos;
+
+$OMERO_DIST/bin/omero config set omero.web.application_server.host 192.168.99.100
```
Comment (joshmoore) [inline home/jobs/OMERO-web/config.xml:91]: This hard-coded IP address won't work elsewhere.

File: home/jobs/OMERO-build/config.xml (hunk 6 above)
Comment (atarkowska) [inline home/jobs/OMERO-build/config.xml:16]: @sbesson suggested that because I was getting timeout. Can be reverted

File: home/jobs/OMERO-web/config.xml (hunk 8 above)
Comment (atarkowska) [inline home/jobs/OMERO-web/config.xml:91]: This is on ToDo list still.

Comment (joshmoore) [inline home/jobs/OMERO-web/config.xml:91]: :+1: Then let's discuss best paths forward tomorrow. My PR isn't working anyway so happy to merge, but you can also take over the PR if you want if that'll simplify the process.

File: home/jobs/OMERO-build/config.xml (hunk 6 above)
Comment (joshmoore) [inline home/jobs/OMERO-build/config.xml:16]: I'll ask. Not sure of the background.

File: docker-compose.yml (hunk 5 above)
Comment (aleksandra-tarkowska) [inline docker-compose.yml]: At this moment I am debating how to dynamically recognise redis host and port in Jenkins job. I was wondering if it would be good idea to introduce Jenkins node to be mounted a volume in another nodes. That would remove @manics hack as well, any thoughts?

Comment (atarkowska) [inline docker-compose.yml:70]: At this moment I am debating how to dynamically recognise redis host and port in Jenkins job. I was wondering if it would be good idea to introduce Jenkins env to be mounted as volume in another nodes to give access to all docker session env. That would remove @manics hack as well, any thoughts?

Comment (joshmoore) [inline docker-compose.yml:70]: You're looking for a way to dynamically recognize them because you want to use different values or because you don't want to hard-code them? In the latter case, you can use `links: redis:redis` and then your `/etc/hosts` file will have a "redis" entry so that you can use the hostname directly.

File: home/jobs/OMERO-build/config.xml (hunk 6 above)
Comment (sbesson) [inline home/jobs/OMERO-build/config.xml:16]: The goal here was to run `OMERO-build` with an existing integration branch i.e. `develop/merge/daily` branch while we were working on setting up the SSH keys. I don't think this is meant to be committed though.

**Comment by atarkowska** [comment]
Clsing in favor of https://github.com/openmicroscopy/devspace/pull/20

File: docker-compose.yml (hunk 5 above)
Comment (atarkowska) [inline docker-compose.yml:70]: but https://github.com/openmicroscopy/devspace/pull/20/files#diff-f9a547ad226d5c0d8302f247b9ff5c39R76 gives 172.17.0.2 instead of 192.168.99.100

Comment (joshmoore) [inline docker-compose.yml:70]: Can you not use the symbolic host name?
